src/sentry/migrations/0866_backfill_open_periods.py

```diff
@@ -0,0 +1,177 @@
+    if status == GroupStatus.UNRESOLVED and len(activities) > 0:
+        open_periods.append(
+            GroupOpenPeriod(
+                group_id=group_id,
+                project_id=project_id,
+                date_started=activities[0].datetime,
+                date_ended=None,
+                resolution_activity=None,
+                user_id=None,
+            )
+        )
+        activities = activities[1:]
+
+    for activity in activities:
+        if activity.type == ActivityType.SET_RESOLVED.value:
+            end = activity.datetime
+            end_activity = activity
+        elif activity.type == ActivityType.SET_REGRESSION.value:
+            start = activity.datetime
+            if start is not None and end is not None:
+                open_periods.append(
+                    GroupOpenPeriod(
+                        group_id=group_id,
+                        project_id=project_id,
+                        date_started=start,
+                        date_ended=end,
+                        resolution_activity=end_activity,
+                        user_id=end_activity.user_id if end_activity else None,
+                    )
+                )
+                end = None
+                end_activity = None
+    # Add the very first open period, which has no UNRESOLVED activity for the group creation
+    open_periods.append(
+        GroupOpenPeriod(
+            group_id=group_id,
+            project_id=project_id,
+            date_started=first_seen,
+            date_ended=end if end else None,
+            resolution_activity=end_activity if end_activity else None,
+            user_id=end_activity.user_id if end_activity else None,
+        )
+    )
+    return open_periods
+
+
+def backfill_group_open_periods(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    Group = apps.get_model("sentry", "Group")
+    GroupOpenPeriod = apps.get_model("sentry", "GroupOpenPeriod")
+
+    batch = []
+
+    for group_id, first_seen, status, project_id in RangeQuerySetWrapperWithProgressBarApprox(
+        Group.objects.all().values_list("id", "first_seen", "status", "project_id"),
+        result_value_getter=lambda item: item[0],
+    ):
+        # Skip groups that already have open periods starting from the first seen date.
+        # These groups were either already backfilled or were created after the open period
+        # logic was added.
+        old_open_period = (
+            GroupOpenPeriod.objects.filter(group_id=group_id).order_by("date_started").first()
+        )
+        if old_open_period and old_open_period.date_started == first_seen:
+            continue
+
+        # Backfill until the first open period that already exists.
+        query_end = old_open_period.date_started if old_open_period else None
+        open_periods = get_open_periods_for_group(
+            apps, group_id, status, project_id, first_seen, query_end, GroupOpenPeriod
+        )
```

Comment:
It might be helpful to batch this too, or at least the activitiy query. We'd have to move the `query_end` filtering to in memory, but that might be worth the speed improvements from not making a query per group here.